### ISSUE TITLE
Accessing REST API

### ISSUE THREAD

**Issue by ayuser123** [LibreBooking/docker#52]
Hi,

Running docker build 2.8.6-2.1.0 (https://hub.docker.com/layers/librebooking/librebooking/2.8.6-2.1.0/images/sha256-32b6cbd1641d6e78b80ca8f259b3051edb5e744f0f83a8b88db4f2d879bdfb73?context=explore), I'm trying to use the REST API to manage librebooking, but my requests are getting an "HTTP 500 Internal Server Error" reply.

Logs show "/var/www/html/Web/Services/.htaccess: Invalid command 'Header', perhaps misspelled or defined by a module not included in the server configuration"

**Comment by colisee** [comment]
Hi @ayuser123,

could you share with me:
1. Your docker-compose.yml file
2. The full logs
3. The way you invoke the REST API (I am not familiar with them)

**Comment by ayuser123** [comment]
1.  yml:
=================================================
version: "3.7"

services:
  db:
    image: linuxserver/mariadb@sha256:9a9d004333c537614fb6ef186e569e44b6a7373fae002f7d2c7e2c15e4ad6b2f
    container_name: librebooking-db
    restart: always
    networks:
      - net
    volumes:
      - vol-db:/config
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/Toronto
      - MYSQL_DATABASE=librebooking
      - MYSQL_ROOT_PASSWORD=password
      - MYSQL_USER=lb_user
      - MYSQL_PASSWORD=lb_user
  app:
    image: librebooking/librebooking:2.8.6-2.1.0
    container_name: librebooking
    restart: always
    depends_on:
      - db
    networks:
      - net
    ports:
      - "8080:80"
    volumes:
      - vol-app:/var/www/html
    environment: 
      - TZ=America/Toronto
      - LB_DB_HOST=db
      - LB_DB_NAME=librebooking
      - LB_INSTALL_PWD=install
      - LB_DB_USER=lb_user
      - LB_DB_USER_PWD=lb_user

volumes:
  vol-db:
    name: librebooking-db_data
  vol-app:
    name: librebooking_html

networks:
  net:
    name: librebooking
=================================================    
    
2. log produced from API request:
172.29.0.1 - - [31/Jul/2023:17:59:31 -0400] "POST /Web/Services/index.php/Authentication/Authenticate/ HTTP/1.1" 500 805 "-" "-"
[Mon Jul 31 17:59:31.328416 2023] [core:alert] [pid 49] [client 172.29.0.1:53546] /var/www/html/Web/Services/.htaccess: Invalid command 'Header', perhaps misspelled or defined by a module not included in the server configuration

3. Request/Reply:
![image](https://github.com/LibreBooking/docker/assets/141073592/229e3b94-de50-4c89-84ed-bf1c47db9264)

**Comment by colisee** [comment]
Hi @ayuser123,

I believe that the module `headers` is not enabled in the image/container.

Could you please try the following command:
```
sudo docker exec -t librebooking bash -c "a2enmod headers && service apache2 restart"
```

If you confirm it works, then I will make the changes and issue a new image release.

**Comment by ayuser123** [comment]
Hi,

Yes enabling the headers in the apache server helped with the 500 error.
Note that I had to edit the config.php file manually to enable the API, it is not enabled by default (and not configurable from the web GUI).

However now the reply is returned with an empty body.
![image](https://github.com/LibreBooking/docker/assets/141073592/b796b600-4cd7-4a7d-a18e-0c3dd1a3f66e)

logs from the terminal show the following error:
[Wed Aug 02 14:31:10.182660 2023] [php:warn] [pid 44] [client 172.29.0.1:44208] PHP Warning:  Undefined array key "request" in /var/www/html/lib/WebService/Slim/SlimServiceRegistration.php on line 120

**Comment by colisee** [comment]
# Work in progress

## Docker project
Bug:
- Add `a2enmod Headers` to Dockerfile

Enhancements:
- Add docker environment `LB_API` to program `entrypoint.sh`. Set default value to `true`
- Ensure that `LB_API` is handled each time `entrypoint.sh` is invoked

## Upstream project
Bug:
- Change line 838 of `lib/Common/Date.php` to `$minutes = intval($this->seconds / 60) % 60;`

**Comment by colisee** [comment]
Hi @ayuser123 ,

I found an issue/bug with the program lib/Common/Data.php as shown above.
Once I fixed the line:
- I was able to display the full page of http://localhost/Web/Services
- I was able to authenticate and got the SessionId and UserID in return

![Capture d’écran 2023-08-03 163957](https://github.com/LibreBooking/docker/assets/64310405/130fd969-8260-47a9-88bb-719687eaa497)

Could you try to make the same program modification inside your running container?

**Comment by ayuser123** [comment]
Hi,

Thanks for all your help with this! 

Local changes summarized: 
- headers enabled.
- API enabled in the confi.php file.
- "LB_API=true" was added to the services/app/environment section of the yml file: [docker-compose.yml.txt](https://github.com/LibreBooking/docker/files/12254429/docker-compose.yml.txt).
- minutes calculation fixed.

I am able to get the /Web/Services description page, but note that this was working for me earlier as well.

Unfortunately, I'm still getting the same "empty" reply body with the "Undefined array key" warning as described earlier.

What build of the docker image are you testing with?

**Comment by colisee** [comment]
Hi,

I am using the image: `librebooking/librebooking:develop-2.1.0`. 
I just confirmed that the authentication does not work with the stable version `librebooking/librebooking:2.8.6-2.1.0`. Probably a fix was brought to the upstream project after the release 2.8.6 was issued.

Here are the steps that I am following:
1. Stop any running instance and delete the existing volumes
   ```
   docker compose down -v
   ```
1. Start a new instance from **scratch**
   ```
   docker compose up -d
   ```
1. Once the containers are ready, initialize the application by pointing your web browser to `http://localhost/Web/install`
1. Enable the headers module
   ```
   docker exec -t librebooking a2enmod headers
   docker exec -t librebooking service apache2 restart
   ```
1. Enable the librebooking api service
   ```
   docker exec -t -u www-data librebooking sed -E 's:(.*api.*enabled.* = .)false:\1true:' -i config/config.php
   ```
1. Fix lib/Common/Date.php
   ```
   docker exec -t -u www-data librebooking sed -i lib/Common/Date.php -e 's:minutes = intval((\$this->seconds / 60) % 60);:minutes = intval(\$this->seconds / 60) % 60;:'
   ```
1. Authenticate with the api (in my case: user = admin and password = LibrebookinG)
   ```
   curl -X POST http://localhost/Web/Services/Authentication/Authenticate -H 'Content-Type: application/json' -d '{"username":"admin","password":"LibrebookinG"}'
   ```

**Comment by ayuser123** [comment]
yes, thanks, we are now able to authenticate on a different docker instance using the above steps! :)  

 "{
    "links": [
        {
            "href": "[http://10.5.196.28:8080/Web/Services/Authentication/SignOut"](http://x.x.x.x:8080/Web/Services/Authentication/SignOut%22),
            "title": "logout"
        }
    ],
    "message": null,
    "sessionToken": "d107894fd92d46821cc641cfdc7c1fc94e80059ad31cccb321",
    "sessionExpires": "2023-08-04T20:56:17+0000",
    "userId": "2",
    "isAuthenticated": true,
    "version": "2.8"
}"

Is there a way to do this without deleting the existing volumes? 
If I destroy only the librebooking_html volume, will the user/reservation/schedules/resources/etc. data be lost?

**Comment by colisee** [comment]
Well done @ayuser123 !

The reason I wanted you to delete the 2 volumes was to make sure the fix worked on a "clean" environment. 

At this stage, I believe that you can keep your 2 volumes, as long as you perform steps 4 through 6 on your running librebooking instance.

This issue will be closed, when I create the docker release 2.1.1 (with the Dockerfile fix). New images will be available on the [docker hub](https://hub.docker.com/r/librebooking/librebooking), but the api won't work since the bug comes from the [upstream project](https://github.com/LibreBooking/app). Eventually, the bug (Date.php) will be fixed in upstream and made available on a future docker image.

**Comment by azamani84** [comment]
@colisee can I kindly get your code querying api please? I've tried both Python and JS but response is empty string, although the response status is 200. I am confused and have no idea what I am doing wrong. I can load localhost/Web/Services and see all endpoints.
Thanks for your help.
